designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,311 @@
+  podSecurityContext:
+    runAsNonRoot: true
+
+  ## Configure the kubeRbacProxy container.
+  kubeRbacProxy:
+    image:
+     registry: gcr.io
+     repository: "kubebuilder/kube-rbac-proxy"
+     tag: v0.13.0
+     pullPolicy: IfNotPresent
+
+  ## Configure the resources.
+  ## More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ 
+  resources:
+    limits:
+      cpu: 500m
+      memory: 128Mi
+    requests:
+      cpu: 5m
+      memory: 64Mi
+
+  ## Configure the controller container.
+  controller:
+    image:
+      registry: docker.io
+      repository: ""
+      tag: latest
+      pullPolicy: IfNotPresent
+
+    ## Configure extra options for liveness probe.
+    ## More info: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes
+    livenessProbe:
+      enabled: true
+      httpGet:
+        path: /healthz
+        port: 8081
+      initialDelaySeconds: 15
+      periodSeconds: 20
+
+    readinessProbe:
+      enabled: true
+      httpGet:
+        path: /readyz
+        port: 8081
+      initialDelaySeconds: 5
+      periodSeconds: 10
+
+    ## Configure the resources.
+    ## More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ 
+    resources:
+      limits:
+        cpu: 500m
+        memory: 128Mi
+      requests:
+        cpu: 10m
+        memory: 64Mi
+
+     ## Configure the rbac roles.
+     serviceAccountName: ""
+
+## Webhook configuration under the `config/webhook` directory
+webhook:
+  ## Whether the `resources` field contains `webhook` in the `config/default/kustomization.yaml` file
+  enabled: true
+  ## More info: https://kubernetes.io/docs/concepts/services-networking/service/
+  service:
+    ports:
+    - name: webhook-server
+      protocol: TCP
+      containerPort: 9443
+
+metrics:
+  enabled: true
+  service:
+    type: ClusterIP
+    ports:
+    - port: 8443
+      protocol: TCP
+
+
+## The domain name of Kubernetes cluster 
+## More info: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/
+clusterDomain: cluster.local
+```
+
+### User Stories
+
+- As a developer, I want to be able to generate a helm chart from a kustomize directory so that I can distribute the helm chart to my users. Also, I want the 
+generation to be as simple as possible without the need to write any additional duplicate files.
+
+- As a user, I want the helm chart can cover all potential configurations when I deploy it on the Kubernetes cluster.
+
+- As a platform engineer, I want to be able to manage different versions and configurations of a project across multiple clusters and environments based on the same distribution artifact (Helm Chart), with versioning and dependency locking for supply chain security.
+
+### Implementation Details/Notes/Constraints [optional]
+
+TBF
```

Comment:
```suggestion

### Implementation Details/Notes/Constraints

#### Plugin Layout

- **Location and Versioning**: The new plugin should follow Kubebuilder standards and
be implemented under `pkg/plugins/optional`. It should be introduced as an alpha version
(`v1alpha`), similar to the [Grafana plugin](https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/optional/grafana/v1alpha).

- **The data should be tracked in PROJECT File**: Usage of the plugin should be tracked in the `PROJECT`
file with the input via flags and options if required. Example entry in the `PROJECT` file:

```yaml
...
plugins:
  helm-chart.go.kubebuilder.io/v1-alpha:
    options:
      <flag/key>: <value>
```

Ensure that user-provided input is properly tracked, similar to how it's done in
other plugins [(see the code in the plugin.go)](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugins/golang/deploy-image/v1alpha1/plugin.go#L58-L75)
and the [(code source to track the data)](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugins/golang/deploy-image/v1alpha1/api.go#L191-L217)
of the deploy-image plugin for reference.

**NOTE** We might not need options/flags in the first implementation. However, we should still track the plugin as 
we do for the Grafana plugin. 

#### Plugin Implementation Structure

Following the structure implementation for the source code of this plugin:

```shell
.
├── helm-chart
│   └── v1alpha1
│       ├── init.go
│       ├── edit.go
│       ├── api.go -> Implement any required change in the scaffolds when we run kubebuilder create api
│       ├── webhook.go -> Implement any required change in the scaffolds when we run kubebuilder create webhook
│       ├── plugin.go
│       └── scaffolds
│           ├── init.go
│           ├── edit.go
│           ├── api.go
│           ├── webhook.go
│           └── internal
│               └── templates
```

**Example**

We might want to have only webhook relate scaffold in the HelmChart when/if
the user run `kubebuilder create webhook`. Therefore, the templates and
logic to change the scaffold in this case would be under this subCommand.

Note that we can inject values and/or overwrite then as we do in other places
using the helpers implemented in Kubebuilder. See for example, that we
only add the NetworkPolices relevant to webhooks when those are created in the
project. For further information check its code implementation [here](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugins/common/kustomize/v2/scaffolds/webhook.go#L93-L146)

#### HelmChart Values Scaffolded by the Plugin

- **Allow values.yaml to be fully re-generated with the flag --force**:

By default, the `values.yaml` file should not
be overwritten. However, users should have the option to overwrite it using
a flag (`--force=true`).

This can be implemented in the specific template as done for other plugins:

```go
if f.Force {
    f.IfExistsAction = machinery.OverwriteFile
} else {
    f.IfExistsAction = machinery.Error
}
```

- **Helm Chart Templates should have conditions**:

Ensure templates install resources based on
conditions defined in the `values.yaml`. Example for CRDs:

```
# To install CRDs
{{- if .Values.crd.install }}
...
{{- end }}
```

- **Customizable Values**: Set customizable values in the `values.yaml`,
  such as defining ServiceAccount names, and whether they should be created or not.
  Furthermore, we should include comments to help end-users understand the source
  of configurations. Example:

```yaml
{{- if .Values.rbac.create }}
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app.kubernetes.io/name: project-v4
    app.kubernetes.io/managed-by: kustomize
  name: {{ .Values.rbac.serviceAccountName }}
  namespace: {{ .Release.Namespace }}
{{- end }}
```

- **Layout of the Helm-Chart**:

Following an example of the expected result of this plugin:

```shell
example-project/
  dist/
  helm-chart/
    example-project-crd/
      ├── Chart.yaml
      ├── templates/
      │   ├── _helpers.tpl
      │   ├── crds/
      │   │   └── <CRDs YAML files generated under config/crds/>
      └── values.yaml
    example-project/
      ├── Chart.yaml
      ├── templates/
      │   ├── _helpers.tpl
      │   ├── crds/
      │   └── <CRDs YAML files generated under config/crds/>
      │   ├── certmanager/
      │   │   └── certificate.yaml
      │   ├── default/
      │   │   └── metrics_service.yaml
      │   ├── manager/
      │   │   └── manager.yaml
      │   │   └── namespace.yaml      
      │   ├── network-policy/
      │   │   ├── allow-metrics-traffic.yaml
      │   │   └── allow-webhook-traffic.yaml // Should be added by the plugin subCommand webhook.go
      │   ├── prometheus/
      │   │   └── monitor.yaml
      │   ├── rbac/
      │   │   ├── kind_editor_role.yaml
      │   │   ├── kind_viewer_role.yaml
      │   │   ├── leader_election_role.yaml
      │   │   ├── leader_election_role_binding.yaml
      │   │   ├── metrics_auth_role.yaml
      │   │   ├── metrics_auth_role_binding.yaml
      │   │   ├── metrics_reader_role.yaml
      │   │   ├── role.yaml
      │   │   ├── role_binding.yaml
      │   │   └── service_account.yaml
      │   ├── samples/
      │   │   └── kind_version_admiral.yaml
      │   ├── webhook/
      │   │   ├── manifests.yaml
      │   │   └── service.yaml
      └── values.yaml
```

- **Example of values.yaml**:

```yaml
namespace:
  create: true
  name: <projectName>-system

# Install CRDs under the template
crd:
  install: false

# Webhook configuration sourced from the `config/webhook`
webhook:
  enabled: true
  conversion:
    enabled: true

## RBAC configuration under the `config/rbac` directory
rbac:
  create: true
  serviceAccountName: "controller-manager"

# Cert-manager configuration
certmanager:
  enabled: false
  issuerName: "letsencrypt-prod"
  commonName: "example.com"
  dnsName: "example.com"

# Network policy configuration sourced from the `config/network_policy`
networkPolicy:
  enabled: false

# Prometheus configuration
prometheus:
  enabled: false

# Sample configuration sourced from the `config/samples`
samples:
  install: true

# Manager configuration sourced from the `config/manager`
manager:
  replicas: 1
  image:
    repository: "controller"
    tag: "latest"
  resources:
    limits:
      cpu: 100m
      memory: 128Mi
    requests:
      cpu: 100m
      memory: 64Mi

# Metrics configuration sourced from the `config/metrics`
metrics:
  enabled: true

# Leader election configuration sourced from the `config/leader_election`
leaderElection:
  enabled: true
  role: "leader-election-role"
  rolebinding: "leader-election-rolebinding"
  

# Controller Manager configuration sourced from the `config/manager`
controllerManager:
  manager:
    args:
    - --metrics-bind-address=:8443
    - --leader-elect
    - --health-probe-bind-address=:8081
    containerSecurityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    image:
      repository: controller
      tag: latest
    resources:
      limits:
        cpu: 500m
        memory: 128Mi
      requests:
        cpu: 10m
        memory: 64Mi
  replicas: 1
  serviceAccount:
    annotations: {}

# Kubernetes cluster domain configuration
kubernetesClusterDomain: cluster.local

# Metrics service configuration sourced from the `config/metrics`
metricsService:
  ports:
  - name: https
    port: 8443
    protocol: TCP
    targetPort: 8443
  type: ClusterIP

# Webhook service configuration sourced from the `config/webhook`
webhookService:
  ports:
  - port: 443
    protocol: TCP
    targetPort: 9443
  type: ClusterIP
```

- **Optional configurations should be disabled by default**:

The HelmChart plugin should not scaffold optional options enabled 
when those are scaffolded as disabled by the default implementation 
of `kustomize/v2` and consequently the `go/v4` plugin used by default. Example:

The dependency on Cert-Manager is disabled by default.

```yaml
From config/default/kusyomization.yaml
# [CERTMANAGER] To enable cert-manager, uncomment all sections with 'CERTMANAGER'. 'WEBHOOK' components are required.
#- ../certmanager
```

Therefore, by default the values.ymal should be scaffold with:

```
# Cert-manager configuration
certmanager:
  enabled: false
```

- **Namespace Creation**:

For Golang projects, the namespace will always be created, and the project will be deployed within it. 
For the Helm chart, we should:

```
# templates/manager/namespace.yaml
{{- if .Values.namespace.create }}
apiVersion: v1
kind: Namespace
metadata:
  name: {{ .Values.namespace.name }}
{{- end }}
```

#### To Sync the changes to the HelmCharts

For all implemementations, we need to check the resources which
are scaffold for each subCommand via the [kustomize](https://kubebuilder.io/plugins/kustomize-v2) plugin
and ensure that we will implement the subCommand of the HelmChart plugin
to the respective scaffolds as well.

##### When kubebuilder create api is executed

We can check in [pkg/plugins/common/kustomize/v2/scaffolds/api.go](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugins/common/kustomize/v2/scaffolds/api.go#L75-L84) 
the boilerplates called and the scaffolds done. Therefore, the 
`pkg/plugins/optional/helm-chart/v1alpha/scaffolds/api.go` 
will also need to have the relevant code implementation to do the same 
scaffolds under `dist/helm-charts/`.

##### When kubebuilder create webhook is executed

In this case, we should add the webhook section to the `values.yaml`
as `enabled:true` and ensure that HelmChart has the relevant templates
for webhooks.

We will need to use the utils helpers such as [AppendCodeIfNotExist](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugin/util/util.go#L101C6-L112)
to achieve this goal. Also, we will need to ensure that we do the equivalent
scaffolds as described above for the create api subCommand.

##### To Sync the Manifests Created with `controller-gen`

Users will need to call the subcommand `edit` passing the plugin to 
ensure that the Helm chart is properly synced.

Therefore, the `PostScaffold` of this command should:

- **Run `make manifests`**: Generate the latest CRDs and other manifests.
- **Copy the files to Helm chart templates**:
  - Copy CRDs: `cp config/crd/bases/*.yaml helm-chart/example-project-crd/templates/crds/`
  - Copy RBAC manifests: `cp config/rbac/*.yaml helm-chart/example-project/templates/rbac/`
  - Copy webhook configurations: `cp config/webhook/*.yaml helm-chart/example-project/templates/webhook/`
  - Copy the manager manifest: `cp config/default/manager.yaml helm-chart/example-project/templates/manager/manager.ymal0`
- **Replace placeholders with Helm values**: Ensure that customized fields, such as the namespace, are properly replaced accordingly.
Example: Replace `name: system` with `{{ .Values.namespace.name }}`.

This ensures the Helm chart is always up-to-date with the latest 
manifests generated by Kubebuilder, maintaining consistency with the 
configured namespace and other customizable fields.

We will need to use the utils helpers such as [ReplaceInFile](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugin/util/util.go#L303-L323)
or [EnsureExistAndReplace](https://github.com/kubernetes-sigs/kubebuilder/blob/c058fb95fe0ccd8d2a3147990251ca501df5eb26/pkg/plugin/util/util.go#L276)
to achieve this goal.
```